codec/codec.go

```diff
@@ -23,6 +23,7 @@ var (
 type Codec interface {
 	MarshalInto(interface{}, *wrappers.Packer) error
 	Unmarshal([]byte, interface{}) error
```

Comment:
Would it make sense to remove `Unmarshal` entirely and have the codec manager do the exhaustion check?

This interface feels weird. (Not really having to do with your PR... but the existing code feels weird)

Reply:
If we'd prefer to leave that as a potential followup - I think that's reasonable too.

Reply:
I think that should be a follow up instead of being part of this PR